openstates/al/bills.py

```diff
@@ -39,222 +43,357 @@ def _categorize_action(action):
     return 'other'
 
 class ALBillScraper(BillScraper):
-
     jurisdiction = 'al'
-
-    def refresh_session(self):
-        url = ('http://alisondb.legislature.state.al.us/acas/ACASLoginFire.asp'
-               '?SESSION=%s') % self.session_id
-        html = self.urlopen(url)
-
-    def scrape(self, chamber, session):
-        self.session_id = self.metadata['session_details'][session]['internal_id']
-        self.base_doc_url = 'http://alisondb.legislature.state.al.us/acas/searchableinstruments/%s/PrintFiles/' % session
-
-        chamber_piece = {'upper': 'Senate',
-                         'lower': 'House+of+Representatives'}[chamber]
-        # resolutions
-        res_url = ('http://alisondb.legislature.state.al.us/acas/SESSResosBySe'
-                   'lectedMatterTransResults.asp?WhichResos=%s&TransCodes='
-                   '{All}&LegDay={All}') % chamber_piece
-        self.scrape_for_bill_type(chamber, session, res_url)
-
-        bill_url = ('http://alisondb.legislature.state.al.us/acas/SESSBillsByS'
-                    'electedMatterTransResults.asp?TransCodes={All}'
-                    '&LegDay={All}&WhichBills=%s') % chamber_piece
-        self.scrape_for_bill_type(chamber, session, bill_url)
-
-
-    def scrape_for_bill_type(self, chamber, session, url):
-
-        self.refresh_session()
-
-        html = self.urlopen(url)
-        doc = lxml.html.fromstring(html)
-
-        # bills are all their own table with cellspacing=4 (skip first)
-        bill_tables = doc.xpath('//table[@cellspacing="4"]')
-        for bt in bill_tables[1:]:
-
-            # each table has 3 rows: detail row, description, blank
-            details, desc, _ = bt.xpath('tr')
-
-            # first <tr> has img, button, sponsor, topic, current house
-            #   current status, committee, committee2, last action
-            tds = details.xpath('td')
-            if len(tds) == 9:
-                # middle _, _ is chamber, last action
-                _, button, sponsor, subject, _, _, com1, com2, _ = tds
-            elif len(tds) == 8:
-                # middle _ is last action
-                _, button, sponsor, subject, _, com1, com2, _ = tds
+    CHAMBERS = {'H': 'lower', 'S': 'upper'}
+    DATE_FORMAT = '%m/%d/%Y'
+
+    # Due to a scrapelib bug, a requests Session is necessary to maintain cookies
+    # Retries are also required since the server periodically returns errors
+    # Once this is fixed, roll back to vanilla scrapelib GETs and POSTs
+    s = requests.Session()
+
+    def sget(self, **request_args):
+        self.log('Special GET request to {}'.format(request_args['url']))
+        retry_counter = 0
+        RETRIES = 5
+        while retry_counter < RETRIES:
+            retry_counter += 1
+            time.sleep(retry_counter ** 2)
```

Comment:
We can wipe this when the scrapelib bug is fixed; until then I think it's clearest to the reader to just keep these four lines instead of adding a mixin.